letsencrypt-nginx/letsencrypt_nginx/tests/configurator_test.py

```diff
@@ -119,35 +150,34 @@ def test_deploy_cert(self):
         access_log = os.path.join(self.work_dir, "access.log")
         error_log = os.path.join(self.work_dir, "error.log")
         self.assertEqual([[['server'],
-                           [['listen', '69.50.225.155:9000'],
+                           [['include', self.config.parser.loc["ssl_options"]],
+                            ['ssl_certificate_key', 'example/key.pem'],
+                            ['ssl_certificate', 'example/fullchain.pem'],
+                            ['error_log', error_log],
+                            ['access_log', access_log],
+
+                            ['listen', '5001 ssl'],
+                            ['listen', '69.50.225.155:9000'],
                             ['listen', '127.0.0.1'],
                             ['server_name', '.example.com'],
-                            ['server_name', 'example.*'],
-                            ['listen', '5001 ssl'],
-                            ['access_log', access_log],
-                            ['error_log', error_log],
-                            ['ssl_certificate', 'example/cert.pem'],
-                            ['ssl_certificate_key', 'example/key.pem'],
-                            ['include',
-                             self.config.parser.loc["ssl_options"]]]]],
+                            ['server_name', 'example.*']]]],
                          parsed_example_conf)
         self.assertEqual([['server_name', 'somename  alias  another.alias']],
                          parsed_server_conf)
-        self.assertEqual([['server'],
-                          [['listen', '8000'],
-                           ['listen', 'somename:8080'],
-                           ['include', 'server.conf'],
-                           [['location', '/'],
-                            [['root', 'html'],
-                             ['index', 'index.html index.htm']]],
-                           ['listen', '5001 ssl'],
-                           ['access_log', access_log],
-                           ['error_log', error_log],
-                           ['ssl_certificate', '/etc/nginx/cert.pem'],
-                           ['ssl_certificate_key', '/etc/nginx/key.pem'],
-                           ['include',
-                            self.config.parser.loc["ssl_options"]]]],
-                         parsed_nginx_conf[-1][-1][-1])
+        self.assertTrue(util.contains_at_depth(parsed_nginx_conf,
```

Comment:
My goal with this change was to make the tests less brittle and more expressive. Test cases that depend very closely on detailed structure tend to break any time there's a change, even if that change did not actually break functionality. This decreases developer confidence in tests, since they frequently produce false positives. Also, "golden" tests that check for exact matches on large chunks of output encourage developers to copy and paste new output to quickly unbreak a test. You can see here that I still did a little of that by simply reordering the server fields instead of breaking down into individual test cases.

Instead, it's nice to have tests that test that a specific thing is not broken. We care that the `ssl_certificate_key` directive is present, and that it has the value `/etc/nginx/key.pem`, but it doesn't matter where in the server block it occurs. So ideally this test would go one step further and loop through each server block, trying to find one that contains each necessary directive, regardless of order. But I didn't want to go too deep in this one change. :-)

Actually, this reminds me, I need to write a test specifically for the new behavior of add_server_directives.